### ISSUE TITLE
`wrap-reader' does not achieve its intended purpose in the Boot toolchain

### ISSUE THREAD

**Issue by danielsz** [Engelberg/instaparse#104]
One of the defining features of the Boot toolchain is that it can accomodate multiple Clojure runtimes. As a consequence, merely checking for the presence of `ConditionalReader` class does not guarantee that the new reader will be available over the course of the build process.
Please consider checking for the version string. 
Offending lines: 
https://github.com/Engelberg/instaparse/blob/d4dff88bc69c05367fc2d4f111b826b6977cf082/src/instaparse/cfg.clj#L159-L163
Reference:
```shell
19:42 < michaniskin> the problem was that two versions of clojure were on the classpath
19:42 < michaniskin> 1.6.0 was on the classpath because that's the default for boot
19:42 < michaniskin> it loads that as the environment in which your build.boot will run
19:43 < michaniskin> then you added 1.7.0-RC1 to :dependencies
19:43 < michaniskin> which loaded a second version of clojure into the classpath
19:43 < michaniskin> this triggers a bug in instaparse where instaparse is sniffing for the presence of a particular class on the classpath as its way of determining which version of clojure it's running in
```

**Comment by Engelberg** [comment]
I don't have Boot installed, so don't have a good way to test a potential patch.  But if you provide a pull-request and can verify it fixes the problem with instaparse in Boot, I will incorporate it.

**Comment by danielsz** [comment]
Thank you. I will consult with Micha Niskin and Alan Dipert (authors of Boot) on this matter.

**Comment by alandipert** [comment]
I would add that the problem can be reproduced without boot thusly:

```
alan@alanputer:~$ java -cp $HOME/.m2/repository/org/clojure/clojure/1.6.0/clojure-1.6.0.jar:\
$HOME/.m2/repository/org/clojure/clojure/1.7.0-RC1/clojure-1.7.0-RC1.jar:\
$HOME/.m2/repository/instaparse/instaparse/1.4.0/instaparse-1.4.0.jar clojure.main
Clojure 1.6.0
user=> (require 'instaparse.core)
CompilerException clojure.lang.ArityException: Wrong number of args (4) passed to: StringReader, compiling:(abnf.clj:186:28)
```

I would add also that I [queried the clojure-dev list](https://groups.google.com/d/msg/clojure-dev/TnLpS7upF4E/jxPifQuflrcJ) about adding a 2-arg StringReader constructor to 1.7 for compatibility but I don't think they're interested.

**Comment by Engelberg** [comment]
I released a 1.4.1 version that uses the version number to handle the string-reader issue.  It still seems like there could be an issue if you do ahead-of-time compilation in one version of Clojure, and then try to use it with another version on the classpath, but this should be an improvement over what is currently there.  Let me know if it works now under boot.

**Comment by alandipert** [comment]
Definitely an improvement, thanks1
